pkg/devfile/resources.go

```diff
@@ -0,0 +1,258 @@
+	}
+
+	deployParams := generator.DeploymentParams{
+		TypeMeta:          generator.GetTypeMeta("Deployment", "apps/v1"),
+		Containers:        containers,
+		PodSelectorLabels: map[string]string{"app": name},
+	}
+
+	deployment, err := generator.GetDeployment(devfileObj, deployParams)
+	if err != nil {
+		return appsv1.Deployment{}, err
+	}
+
+	return *deployment, nil
+}
+
+// GetService gets the service resource
+func GetService(devfileObj parser.DevfileObj, filterOptions common.DevfileOptions, imagePort string) (corev1.Service, error) {
+
+	serviceParams := generator.ServiceParams{
+		TypeMeta: generator.GetTypeMeta("Service", "v1"),
+	}
+
+	service, err := generator.GetService(devfileObj, serviceParams, filterOptions)
+	if err != nil {
+		return corev1.Service{}, err
+	}
+
+	portNumber, err := strconv.Atoi(imagePort)
+	if err != nil {
+		return corev1.Service{}, err
+	}
+
+	svcPort := corev1.ServicePort{
+		Name:       fmt.Sprintf("http-%v", imagePort),
+		Port:       int32(portNumber),
+		TargetPort: intstr.FromString(imagePort),
+	}
+	service.Spec.Ports = append(service.Spec.Ports, svcPort)
+
+	return *service, nil
+}
+
+// GetRouteForDockerImage gets the route resource
+func GetRouteForDockerImage(name string, port, path string, secure bool, annotations map[string]string) routev1.Route {
+
+	if path == "" {
+		path = "/"
+	}
+
+	routeParams := generator.RouteParams{
+		TypeMeta: generator.GetTypeMeta("Route", "route.openshift.io/v1"),
+		RouteSpecParams: generator.RouteSpecParams{
+			ServiceName: name,
+			PortNumber:  intstr.FromString(port),
+			Path:        path,
+			Secure:      secure,
+		},
+	}
+
+	return *generator.GetRoute(devfilev1.Endpoint{Annotations: annotations}, routeParams)
+}
+
+// GetResourceFromDevfile gets the deployment, service and route resources from the devfile
+func GetResourceFromDevfile(devfileObj parser.DevfileObj, deployAssociatedComponents map[string]string, name string) (appsv1.Deployment, corev1.Service, routev1.Route, error) {
+	kubernetesComponentFilter := common.DevfileOptions{
+		ComponentOptions: common.ComponentOptions{
+			ComponentType: devfilev1.KubernetesComponentType,
+		},
+	}
+	kubernetesComponents, err := devfileObj.Data.GetComponents(kubernetesComponentFilter)
+	if err != nil {
+		errMsg := fmt.Sprintf("Failed to get the kubernetes components from devfile: %v", err)
+		klog.Error(errMsg)
+		return appsv1.Deployment{}, corev1.Service{}, routev1.Route{}, err
+	}
+
+	var appendedResources parser.KubernetesResources
+	var deployment appsv1.Deployment
+	var service corev1.Service
+	var route routev1.Route
+
+	returnDeploymentOnly := true // NOTE: (TODO) once devfile samples YAML have service & route definitions, switch to false
+	// Update tests accordingly
+
+	for _, component := range kubernetesComponents {
+		if _, ok := deployAssociatedComponents[component.Name]; ok && component.Kubernetes != nil {
+			var src parser.YamlSrc
+
+			if component.Kubernetes.Inlined != "" {
```

Comment:
instead of a if-else statement. can check `if component.Kubernetes.Inlined == ""`, and do the dev preview way. since it's going to break out if falls in that if statement. the `src` definition can be defined with all the stuffs starting at [line 186](https://github.com/openshift/console/blob/f1d0fd8d641b0dbb6810681be25de717624f7ca7/pkg/devfile/resources.go#L186)